### PR TITLE
Parameterize over what it means to be an Asset (Coin v.s. Value)

### PR DESCRIPTION
**Pull request by TimSheard**
This PR parameterizes over what kind of assets a transaction can produce. In the old style we used a Coin, and as an alternative to that we introduce the type Value (which is a kind of multi-asset). Both these types are instances of the Val class, so the code has been made parametric over objects in the Val class.

### CODE REVIEW

File: shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/Genesis.hs [hunk 1]
```diff
@@ -102,19 +102,20 @@ data ShelleyGenesis c = ShelleyGenesis
...
 
-sgActiveSlotCoeff :: ShelleyGenesis c -> ActiveSlotCoeff
+sgActiveSlotCoeff :: ShelleyGenesis c v -> ActiveSlotCoeff
 sgActiveSlotCoeff =
   mkActiveSlotCoeff
     . unitIntervalFromRational
     . sgActiveSlotsCoeff
 
-instance Crypto crypto => ToJSON (ShelleyGenesis crypto) where
+-- TODO JSON for v?
```
Comment (JaredCorduan) [inline shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/Genesis.hs:117]: I think we can remove this TODO

File: shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/LedgerState.hs [hunk 2]
```diff
@@ -633,65 +634,59 @@ genesisState genDelegs0 utxo0 =
...
 -- | Implementation of abstract transaction size
-txsize :: Tx crypto -> Integer
+txsize :: Tx crypto v -> Integer
 txsize = fromIntegral . BSL.length . txFullBytes
 
 -- | Convenience Function to bound the txsize function.
 -- | It can be helpful for coin selection.
-txsizeBound :: forall crypto. (Crypto crypto) => Tx crypto -> Integer
+txsizeBound :: forall crypto v. (CV crypto v) => Tx crypto v -> Integer
 txsizeBound tx = numInputs * inputSize + numOutputs * outputSize + rest
```
Comment (JaredCorduan) [inline shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/LedgerState.hs:643]: I think we might be able to get rid of this function altogether. which would mean we could also clean up the transaction type (it stores the input and output bytes so that `txsizeBound` can remove them). we can do this in a separate PR.

File: shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/LedgerState.hs [hunk 3]
```diff
@@ -633,65 +634,59 @@ genesisState genDelegs0 utxo0 =
...
+  CV crypto v =>
   PParams ->
-  TxBody crypto ->
+  TxBody crypto v ->
   Coin
 keyRefunds pp tx = (_keyDeposit pp) * (fromIntegral $ length deregistrations)
   where
     deregistrations = filter isDeRegKey (toList $ _certs tx)
 
+-- TODO add forge
```
Comment (JaredCorduan) [inline shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/LedgerState.hs:679]: I don't think we need to add all these TODOs in the code

Comment (JaredCorduan) [reply shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/LedgerState.hs:679]: a github issue would be more helpful

File: shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/STS/Utxo.hs [hunk 4]
```diff
@@ -159,20 +159,20 @@ instance
       PostCondition
         "Deposit pot must not be negative (post)"
         (\_ st' -> _deposited st' >= 0),
-      let utxoBalance us = _deposited us + _fees us + balance (_utxo us)
+      let utxoBalance us = vplus (vinject $ _deposited us + _fees us) (balance (_utxo us))
```
Comment (JaredCorduan) [inline shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/STS/Utxo.hs:162]: can we use `<>` to make it a bit more like the original?

File: shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/STS/Utxo.hs [hunk 5]
```diff
@@ -301,11 +304,18 @@ utxoInductive = do
   -- process Protocol Parameter Update Proposals
   ppup' <- trans @(PPUP crypto) $ TRC (PPUPEnv slot pp genDelegs, ppup, txup tx)
 
-  let outputs = Set.toList (eval (rng (txouts txb)))
+  -- TODO check voper
```
Comment (JaredCorduan) [inline shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/STS/Utxo.hs:307]: can we remove this? or do we need to check something?

File: shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/STS/Utxo.hs [hunk 6]
```diff
@@ -301,11 +304,18 @@ utxoInductive = do
...
       minUTxOValue = _minUTxOValue pp
-      outputsTooSmall = [out | out@(TxOut _ c) <- outputs, c < minUTxOValue]
+      outputsTooSmall = [out | out@(TxOut _ vl) <- outputs, (voper Gt) (vinject $ scaleVl vl minUTxOValue) vl]
   null outputsTooSmall ?! OutputTooSmallUTxO outputsTooSmall
 
+  -- TODO add forge
+  -- TODO add forge errors
+  -- let (Value vls) = _forge txb
+  -- let cids = Map.keys vls
+  -- all (adaID /=) cids  ?! (ForgingAda (Value vls))
```
Comment (JaredCorduan) [inline shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/STS/Utxo.hs:317]: let's remove this and make an issue instead

File: shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/Tx.hs [hunk 7]
```diff
@@ -109,51 +110,44 @@ import Shelley.Spec.Ledger.TxData
...
+deriving instance Crypto crypto => Show (WitnessSetHKD Identity crypto v)
 
-deriving instance Crypto crypto => Eq (WitnessSetHKD Identity crypto)
+deriving instance Crypto crypto => Eq (WitnessSetHKD Identity crypto v)
 
-deriving instance Crypto crypto => Generic (WitnessSetHKD Identity crypto)
-
-deriving via
-  AllowThunksIn
-    '[ "txWitsBytes"
```
Comment (JaredCorduan) [inline shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/Tx.hs:133]: don't we still want the annotations to be thunks?

File: shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/STS/Utxo.hs (hunk 6 above)
Comment (polinavino) [reply shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/STS/Utxo.hs:317]: This is more to do with actually adding forging functionality than just doing the parametrization. Can definitely be removed from here.

File: shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/Tx.hs (hunk 7 above)
Comment (polinavino) [reply shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/Tx.hs:133]: We couldnt get deriving via to work with `AllowThunksIn` because of lack of ability to derive `Generic` for everything in `Val`. Is this essential?

File: shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/TxData.hs [hunk 8]
```diff
@@ -403,40 +452,46 @@ pattern TxIn addr index <-
...
+deriving via UseIsNormalFormNamed "TxIn" (TxIn crypto v) instance NoUnexpectedThunks (TxIn crypto v)
+
+-- | Parametrized tx output
+-- TODO make v compact too
+data TxOut crypto v where
+  TxOutCompact ::
+    CV crypto v =>
+    {-# UNPACK #-} !BSS.ShortByteString ->
+    v ->
+    TxOut crypto v
```
Comment (JaredCorduan) [inline shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/TxData.hs:464]: it seems we can no longer unpack the Word64 from the coin if we want this to be parametric, so the UTxO memory size will be a bit bigger (one word per TxOut?). we can maybe make an issue to measure the difference.

File: shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/TxData.hs [hunk 9]
```diff
@@ -534,24 +593,29 @@ data TxBody crypto = TxBody'
...
   }
-  deriving (Show, Eq, Generic)
-  deriving
-    (NoUnexpectedThunks)
-    via AllowThunksIn '["bodyBytes"] (TxBody crypto)
 
-instance Crypto c => HashAnnotated (TxBody c) c
+instance CV c v => HashAnnotated (TxBody c v) c
+
+-- TODO no thunks in bodyBytes is ok?
```
Comment (JaredCorduan) [inline shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/TxData.hs:599]: that's a good point, I think we want the annotations to be thunks

File: shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/TxData.hs [hunk 10]
```diff
@@ -754,37 +818,32 @@ instance
...
   fromCBOR = decodeRecordNamed "TxOut" (const 2) $ do
-    bs <- fromCBOR
-    coin <- fromCBOR
-    -- Check that the address is valid by decompacting it instead of decoding
-    -- it as an address, as that would require compacting (re-encoding) it
-    -- afterwards.
-    case decompactAddr bs of
-      Just (_ :: Addr crypto) -> pure $ TxOutCompact bs coin
-      Nothing -> cborError $ DecoderErrorCustom "TxOut" "invalid address"
+    addr <- fromCBOR
```
Comment (JaredCorduan) [inline shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/TxData.hs:834]: I think we still need to check that the address parses with `decompactAddr`

File: shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/UTxO.hs [hunk 11]
```diff
@@ -163,64 +163,64 @@ txouts tx =
...
+  Set (WitVKey crypto v ('Witness))
 makeWitnessesFromScriptKeys txbodyHash hashKeyMap scriptHashes =
   let witKeys = Map.restrictKeys hashKeyMap scriptHashes
    in makeWitnessesVKey txbodyHash (Map.elems witKeys)
 
 -- | Determine the total balance contained in the UTxO.
-balance :: UTxO crypto -> Coin
-balance (UTxO utxo) = fromIntegral $ Map.foldl' addCoins 0 utxo
+balance :: CV crypto v => UTxO crypto v -> v
+balance (UTxO utxo) = foldr addCoins vzero utxo
```
Comment (JaredCorduan) [inline shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/UTxO.hs:221]: `foldl'`?

File: shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/UTxO.hs [hunk 12]
```diff
@@ -163,64 +163,64 @@ txouts tx =
...
   where
-    addCoins !b (TxOutCompact _ a) = a + b
+    addCoins (TxOut _ a) b = vplus a b
```
Comment (JaredCorduan) [inline shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/UTxO.hs:223]: I think we still want the real constructor `TxOutCompact`. and maybe `<>` would be more clear

File: shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/UTxO.hs [hunk 13]
```diff
@@ -291,14 +292,32 @@ scriptsNeeded u tx =
...
+  where
+    outputsUsingScripts =
+      [out | out@(TxOut addr _) <- Map.elems $ unUTxO (txouts txb), addrUsesScript addr]
+    stakeAddrCertsUsingScripts =
+      [cert | cert@(DCertDeleg (RegKey sc)) <- toList (_certs txb), credentialUsesScript sc]
+    poolRegCertsUsingScripts =
+      [cert | cert@(DCertPool (RegPool pparams)) <- toList (_certs txb), rewardAcntUsesScript (_poolRAcnt pparams)]
+    mirCertsUsingScripts =
+      [cert | cert@(DCertMir mir) <- toList (_certs txb), any credentialUsesScript (Map.keys (mirRewards mir))]
+-}
```
Comment (JaredCorduan) [inline shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/UTxO.hs:323]: dead code?

File: shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/UTxO.hs (hunk 12 above)
Comment (mrBliss) [reply shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/UTxO.hs:223]: Yeah, please don't undo https://github.com/input-output-hk/cardano-ledger-specs/pull/1771. 

(Maybe you started working on this with an out-of-date checkout, one that didn't include that PR, and this change got lost while merging/rebasing? If so, we should be careful that no other such changes are in this PR.)

File: shelley/chain-and-ledger/shelley-spec-ledger-test/src/Test/Shelley/Spec/Ledger/Generator/Constants.hs [hunk 14]
```diff
@@ -88,51 +88,54 @@ data Constants = Constants
...
+    maxMinFeeA = 1000,
+    maxMinFeeB = 3,
+    numCoreNodes = 7,
+    minTreasury = 1000000,
+    maxTreasury = 10000000,
+    minReserves = 1000000,
+    maxReserves = 10000000,
+    genTxRetries = 6,
+    genTxBudget = 10000
+  }
```
Comment (JaredCorduan) [inline shelley/chain-and-ledger/shelley-spec-ledger-test/src/Test/Shelley/Spec/Ledger/Generator/Constants.hs:141]: let's not indent so that we can see exactly what changed

File: shelley/chain-and-ledger/shelley-spec-ledger-test/src/Test/Shelley/Spec/Ledger/Generator/Core.hs [hunk 15]
```diff
@@ -58,7 +58,8 @@ import Cardano.Crypto.DSIGN.Class (DSIGNAlgorithm (..))
 import Cardano.Crypto.VRF (evalCertified)
 import qualified Cardano.Crypto.Hash as Hash
 import Control.Iterate.SetAlgebra (eval, (∪), (⋪))
-import Control.Monad (replicateM)
+
```
Comment (JaredCorduan) [inline shelley/chain-and-ledger/shelley-spec-ledger-test/src/Test/Shelley/Spec/Ledger/Generator/Core.hs:62]: mystery line! :)

File: shelley/chain-and-ledger/shelley-spec-ledger-test/src/Test/Shelley/Spec/Ledger/Generator/MetaData.hs [hunk 16]
```diff
@@ -32,10 +32,11 @@ genMetaData ::
...
 genMetaData (Constants {frequencyTxWithMetaData}) =
-  QC.frequency
-    [ (frequencyTxWithMetaData, genMetaData'),
-      (100 - frequencyTxWithMetaData, pure (SNothing, SNothing))
-    ]
+  do (a,b) <- QC.frequency
+                [ (frequencyTxWithMetaData, genMetaData'),
+                  (100 - frequencyTxWithMetaData, pure (SNothing, SNothing))
+                ]
+     pure(a,b)
```
Comment (JaredCorduan) [inline shelley/chain-and-ledger/shelley-spec-ledger-test/src/Test/Shelley/Spec/Ledger/Generator/MetaData.hs:39]: I think this doesn't actually change anything

File: shelley/chain-and-ledger/shelley-spec-ledger-test/src/Test/Shelley/Spec/Ledger/Generator/Trace/DCert.hs [hunk 17]
```diff
@@ -150,7 +150,7 @@ genDCerts ::
       ([KeyPair 'Witness c], [(MultiSig c, MultiSig c)])
     )
 genDCerts
-  ge@( GenEnv
+   ge@( GenEnv
```
Comment (JaredCorduan) [inline shelley/chain-and-ledger/shelley-spec-ledger-test/src/Test/Shelley/Spec/Ledger/Generator/Trace/DCert.hs:153]: we don't need this

File: shelley/chain-and-ledger/shelley-spec-ledger-test/src/Test/Shelley/Spec/Ledger/Generator/Update.hs [hunk 18]
```diff
@@ -298,11 +298,12 @@ genUpdate
...
-                ]
+           in do (a,b) <- QC.frequency
+                           [ ( frequencyTxUpdates,
+                              (,wits) <$> genUpdateForNodes c s e genesisKeys pp
+                              ),
+                            ( 100 - frequencyTxUpdates,
+                             pure (Nothing, [])
+                            )
+                           ]
+                 pure(a,b)
```
Comment (JaredCorduan) [inline shelley/chain-and-ledger/shelley-spec-ledger-test/src/Test/Shelley/Spec/Ledger/Generator/Update.hs:309]: also not needed

File: shelley/chain-and-ledger/shelley-spec-ledger-test/src/Test/Shelley/Spec/Ledger/Generator/Utxo.hs [hunk 19]
```diff
@@ -57,18 +62,18 @@ import Shelley.Spec.Ledger.LedgerState
     DState (..),
     KeyPairs,
     UTxOState (..),
-    minfeeBound,
+    minfee,
```
Comment (JaredCorduan) [inline shelley/chain-and-ledger/shelley-spec-ledger-test/src/Test/Shelley/Spec/Ledger/Generator/Utxo.hs:65]: :tada:

File: shelley/chain-and-ledger/shelley-spec-ledger-test/src/Test/Shelley/Spec/Ledger/Generator/Utxo.hs [hunk 20]
```diff
@@ -82,6 +87,8 @@ import Shelley.Spec.Ledger.UTxO
     makeWitnessesFromScriptKeys,
     makeWitnessesVKey,
   )
+-- import Shelley.Spec.Ledger.Value (Val (..))
```
Comment (JaredCorduan) [inline shelley/chain-and-ledger/shelley-spec-ledger-test/src/Test/Shelley/Spec/Ledger/Generator/Utxo.hs:90]: can remove

File: shelley/chain-and-ledger/shelley-spec-ledger-test/src/Test/Shelley/Spec/Ledger/Generator/Utxo.hs [hunk 21]
```diff
@@ -173,40 +179,219 @@ genTxRetry
       -- SpendingBalance, Output Addresses (including some Pointer addresses)
       -- and a Outputs builder that distributes the given balance over addresses.
       -------------------------------------------------------------------------
-      let spendingBalance = spendingBalanceUtxo + (sum (snd <$> wdrls)) - deposits + refunds
+      let spendingBalance = vplus spendingBalanceUtxo (vinject $ (sum (snd <$> wdrls)) - deposits + refunds)
+          n = if (Map.size . unUTxO) utxo < 100 then 3 else 0
```
Comment (JaredCorduan) [inline shelley/chain-and-ledger/shelley-spec-ledger-test/src/Test/Shelley/Spec/Ledger/Generator/Utxo.hs:183]: we could add this `100` to the constants or reuse another constant...

File: shelley/chain-and-ledger/shelley-spec-ledger-test/src/Test/Shelley/Spec/Ledger/Generator/Utxo.hs [hunk 22]
```diff
@@ -359,18 +545,20 @@ genWithdrawals
...
+            pure ([], ([], []))
+          ),
+          ( frequencyAFewWithdrawals,
+            genWrdls (take maxAFewWithdrawals . Map.toList $ wdrls)
+          ),
+          ( frequencyPotentiallyManyWithdrawals,
+            genWrdls (Map.toList wdrls)
+          )
+        ]
+    pure ( a, b)
```
Comment (JaredCorduan) [inline shelley/chain-and-ledger/shelley-spec-ledger-test/src/Test/Shelley/Spec/Ledger/Generator/Utxo.hs:561]: another one that didn't change?

File: shelley/chain-and-ledger/shelley-spec-ledger-test/src/Test/Shelley/Spec/Ledger/Serialisation/Generators.hs [hunk 23]
```diff
@@ -248,17 +252,16 @@ sizedMetaDatum 0 =
...
+    oneof
+      [ MD.Map <$>
+          (zip
+            <$> (resize maxMetaDatumListLens (listOf (sizedMetaDatum (n-1))))
+            <*> (listOf (sizedMetaDatum (n-1)))),
+        MD.List <$> resize maxMetaDatumListLens (listOf (sizedMetaDatum (n-1))),
+        MD.I <$> arbitrary,
+        MD.B <$> arbitrary,
+        MD.S <$> (T.pack <$> arbitrary)
+      ]
```
Comment (JaredCorduan) [inline shelley/chain-and-ledger/shelley-spec-ledger-test/src/Test/Shelley/Spec/Ledger/Serialisation/Generators.hs:264]: just indendation?

File: shelley/chain-and-ledger/shelley-spec-ledger-test/src/Test/Shelley/Spec/Ledger/Serialisation/Generators.hs [hunk 24]
```diff
@@ -625,12 +628,12 @@ maxMultiSigListLens = 5
...
-      RequireAllOf <$> resize maxMultiSigListLens (listOf (sizedMultiSig (n -1))),
-      RequireAnyOf <$> resize maxMultiSigListLens (listOf (sizedMultiSig (n -1))),
-      RequireMOf <$> arbitrary <*> resize maxMultiSigListLens (listOf (sizedMultiSig (n -1)))
-    ]
+    oneof
+      [ RequireSignature <$> arbitrary,
+        RequireAllOf <$> resize maxMultiSigListLens (listOf (sizedMultiSig (n-1))),
+        RequireAnyOf <$> resize maxMultiSigListLens (listOf (sizedMultiSig (n-1))),
+        RequireMOf <$> arbitrary <*> resize maxMultiSigListLens (listOf (sizedMultiSig (n-1)))
+      ]
```
Comment (JaredCorduan) [inline shelley/chain-and-ledger/shelley-spec-ledger-test/src/Test/Shelley/Spec/Ledger/Serialisation/Generators.hs:636]: just indentation

File: shelley/chain-and-ledger/shelley-spec-ledger-test/src/Test/Shelley/Spec/Ledger/Serialisation/Generators/Genesis.hs [hunk 25]
```diff
@@ -43,9 +44,16 @@ import Shelley.Spec.Ledger.Keys
 import Shelley.Spec.Ledger.PParams
 import Shelley.Spec.Ledger.Scripts
 import Shelley.Spec.Ledger.TxData
+import Shelley.Spec.Ledger.Value
 import Test.Shelley.Spec.Ledger.Utils (mkHash)
 
-genShelleyGenesis :: Crypto c => Gen (ShelleyGenesis c)
+-- TODO this is temporary until I look through the Hedgehog docs
```
Comment (JaredCorduan) [inline shelley/chain-and-ledger/shelley-spec-ledger-test/src/Test/Shelley/Spec/Ledger/Serialisation/Generators/Genesis.hs:50]: do we need this now?

**Review by JaredCorduan — COMMENTED** [review]
Thank you for this massive undertaking!! All in all it looks good to me, but I left several comments, but it's all small stuff.

File: shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/Tx.hs (hunk 7 above)
Comment (redxaxder) [reply shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/Tx.hs:133]: If the field is lazy but the instance doesn't mark it as allowed, this will trip thunk detection.

If the field is made strict, this will trigger serialization every time a witness set is modified.

One solution would be to figure out a way to get a generic instance for the val. Another would be to write a manual instance.
